Add PLEP 8 on switching to calendar versioning

namurphy opened the pull request:
This PR follows up on a discussion in #38 and elsewhere in the plasmapythoniverse about switching from semantic versioning to date-based versioning.  There's a discussion of the tradeoffs in #38.  

I'm basing the scheme in the first draft on Ubuntu's, which has versions like ``22.04`` and ``22.10``, which were the releases from April and October 2022.  The year format is concise, and it's more clear with the zero-padded month that it actually represents a month.  (I got confused earlier today with pip ``v21.3``, since I thought that the minor version number meant the month when it actually didn't.)

This new PLEP would repeal PLEP 5, but since it's a complete rewrite, I decided on doing a new PLEP so as to make the history more apparent.

Closes #38.

namurphy commented on PLEP-0008.rst:
```diff
@@ -0,0 +1,113 @@
+PlasmaPy final releases will use `calendar based versioning`_ of the
+form ``YY.0M.MICRO``, where:
+
+* The major release number ``YY`` is the year minus 2000,
```
Would this be better with a two-digit year or a four-digit year?

StanczakDominik replied:
4-digit YYYY, for the same reasons as https://github.com/dask/community/issues/100#issuecomment-704499489 :)

StanczakDominik commented on PLEP-0008.rst:
```diff
@@ -0,0 +1,113 @@
+that there will be some backward incompatible changes during every
+release. Hence, date-based versioning is more appropriate to PlasmaPy
+than versioning based on whether or not backwards incompatible changes
+have been made.
```
"Because a community software project will have features"... ? I think I get what you were going for here, but can you rephrase?

rocco8773 commented on README.rst:
```diff
@@ -38,3 +38,6 @@ convenience, an abbreviated index of PLEPs is provided below.
 |        |                                                                                  |    :target:  http://doi.org/10.5281/zenodo.3774573                 |
 +--------+----------------------------------------------------------------------------------+--------------------------------------------------------------------+
+| 8      | `Calendar Versioning <./PLEP-0008.rst>`__                                        |                                                                    |
+|        |                                                                                  |                                                                    |
```
Do you plan to have a followup PR to put the doi in?  Or, will that be an update to this PR?

rocco8773 replied:
Ping... Do we have a DOI to put here now?

rocco8773 commented on PLEP-0008.rst:
```diff
@@ -0,0 +1,166 @@
+Versioning Specification
+------------------------
+
+PlasmaPy final releases will use `calendar versioning`_ of the
+form ``YYYY.0M.MICRO``, where:
```
Leading zeros will be removed in accordance with [PEP440](https://peps.python.org/pep-0440/#integer-normalization).  Basically these are normalized in accordance with `int()`.

rocco8773 replied:
```suggestion
form ``YYYY.M.MICRO``, where:
```

Leading zeros will be removed in accordance with [PEP440](https://peps.python.org/pep-0440/#integer-normalization).  Basically these are normalized in accordance with `int()`.

rocco8773 commented on PLEP-0008.rst:
```diff
@@ -0,0 +1,166 @@
+===============================
+
++-------------------+---------------------------------------------+
+| PLEP              | number                                      |
```
```suggestion
| PLEP              | 8                                           |
```

rocco8773 commented on PLEP-0008.rst:
```diff
@@ -0,0 +1,166 @@
++-------------------+---------------------------------------------+
+| date created      | 2022-11-22                                  |
++-------------------+---------------------------------------------+
+| date last revised | 2022-11-29                                  |
```
```suggestion
| date last revised | 2022-12-08                                  |
```

[optional ... ?] just update to current date

rocco8773 commented on PLEP-0008.rst:
```diff
@@ -0,0 +1,166 @@
++-------------------+---------------------------------------------+
+| status            | discussion                                  |
++-------------------+---------------------------------------------+
+| DOI               |                                             |
```
update with reserved DOI

rocco8773 commented on PLEP-0008.rst:
```diff
@@ -0,0 +1,166 @@
+form ``YYYY.0M.MICRO``, where:
+
+* The major release number ``YYYY`` is the year,
+* The minor release number ``0M`` is the zero-padded month, and
```
```suggestion
* The minor release number ``M`` is positive integer representing the release month, and
```

See above PEP440 comment.  Feel to re-word to your own liking.

rocco8773 commented on PLEP-0008.rst:
```diff
@@ -0,0 +1,166 @@
+* The bugfix release number ``MICRO`` will start from ``0`` and be
+  incremented by one for each bugfix release.
+
+For example, the first release in April 2023 would be ``2023.04.0``,
+with ``2023.04.1`` as the first bugfix release of the ``2023.04.*``
+series.
```
```suggestion
For example, the first release in April 2023 would be ``2023.4.0``,
with ``2023.4.1`` as the first bugfix release of the ``2023.4.*``
series.
```

Again, refer to the above PEP440 comments.

rocco8773 commented on PLEP-0008.rst:
```diff
@@ -0,0 +1,166 @@
+  roughly proportional to the time between releases, and can therefore
+  be gauged more accurately with date-based versioning.
+
+* The versioning scheme would be consistent with projects that use some
+  form of calendar versioning such as xarray, dask, pip, Ubuntu, and
+  PyCharm.
```
```suggestion
* The versioning scheme would be consistent with projects that use some
  form of calendar versioning such as ``xarray``, ``dask``, ``pip``,
  Ubuntu, and PyCharm.
```

[optional] ... but thank you for using the Oxford comma. 😺

rocco8773 commented on PLEP-0008.rst:
```diff
@@ -0,0 +1,166 @@
+Advantages of adopting calendar versioning
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+* The age of a release can be gauged directly from the version number.
```
Not sure the best way to make the comment, but also easier to traceback the version when cited in a journal article.

namurphy commented on PLEP-0008.rst:
```diff
@@ -0,0 +1,166 @@
+
+Additional considerations for adopting calendar versioning
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+* A change to calendar versioning would need to be permanent in order to
+  preserve version monotonicity.
```
```suggestion
```
There's something in PEP 440 about defining epochs, so this change need not be permanent.

rocco8773 commented on PLEP-0008.rst:
```diff
@@ -0,0 +1,163 @@
+* Calendar versioning is compatible with :pep:`440` and
+  ``setuptools_scm``.
+
+* Using the full year in the version number will reduce the chance of
+  ambiguity. Similarly, using the zero-padded month will increase the
+  likelihood that the minor version number is interpreted as a month.
```
```suggestion
* Using the full year in the version number will reduce the chance of
  ambiguity. Similarly.
```

I'm just deleting this, but feel free to add back a comment about the month but without the "zero-padded" statement.

StanczakDominik commented on PLEP-0008.rst:
```diff
@@ -36,15 +36,15 @@ Versioning Specification
 
 * The major release number ``YYYY`` is the year,
-* The minor release number ``0M`` is the zero-padded month, and
+* The minor release number ``MM`` is positive integer representing the release month, and
```
```suggestion
* The minor release number ``MM`` is a positive integer representing the release month, and
```

StanczakDominik commented:
I'd just like to bring https://github.com/dask/community/issues/100#issuecomment-704468187 to your attention on the zero-padding discussion! The important points:

> 20.01.01 / 2020.01.01 are valid PEP 440 versions.

> Overall, the choice between 0Y/0M/0D vs YY/MM/DD is purely a aesthetic one -- either will work with basically every Python Packaging tool that follows the standards.

I'm weakly in favor of zero-padding, but don't mind much either way.

namurphy commented:
I decided against discussing our LTS strategy since we don't have plans to do that in the near future, and I'd rather have a separate PLEP on that which we can do when we decide to do the LTS releases.

Other than that, I think I addressed all suggestions.  @rocco8773 — if this looks good to you, please feel free to merge.

namurphy commented on PLEP-0005.rst:
```diff
@@ -16,12 +16,17 @@ PLEP-0005 – PlasmaPy Versioning and Releases
 |                   | <https://doi.org/10.5281/zenodo.1451974>`__ |
 +-------------------+---------------------------------------------+
 
+.. important::
+
+    This PLEP has been superseded by |PLEP 8| and is therefore no longer
+    in effect.
```
```suggestion
This PLEP has been superseded by |PLEP 8| and is therefore no longer
in effect.
```

namurphy commented on PLEP-0000.rst:
```diff
@@ -2,81 +2,116 @@
+
++--------+-------------------------------------+------------------------+------------+---------------------------+
+| number | title                               | author(s)              | type       | DOI                       |
++========+=====================================+========================+============+===========================+
+| 5      | `PlasmaPy Versioning and Releases`_ | Nicholas A. Murphy and | process    | `10.5281/zenodo.1451974`_ |
+|        |                                     | Stuart J. Mumford      |            |                           |
++--------+--------------------------------------+------------------------+-----------+---------------------------+
```
Need to fix table!

rocco8773 commented on PLEP-0005.rst:
```diff
@@ -257,3 +260,5 @@ Semantic versioning is a well-defined versioning scheme that provides
 backward incompatible changes.  This scheme is well-suited to a core
 scientific software package that will require stability.
+
+.. |PLEP 8| replace:: https://github.com/PlasmaPy/PlasmaPy-PLEPs/blob/main/PLEP-0007.rst
```
```suggestion
.. |PLEP 7| replace:: https://github.com/PlasmaPy/PlasmaPy-PLEPs/blob/main/PLEP-0007.rst
```

rocco8773 commented on README.rst:
```diff
@@ -38,3 +38,6 @@ convenience, an abbreviated index of PLEPs is provided below.
 | 7      | `Structure of Top-Level Sub-Packages <./PLEP-0007.rst>`__                        | .. image:: https://zenodo.org/badge/DOI/10.5281/zenodo.3774573.svg |
 |        |                                                                                  |    :target:  http://doi.org/10.5281/zenodo.3774573                 |
 +--------+----------------------------------------------------------------------------------+--------------------------------------------------------------------+
+| 8      | `Calendar Versioning <./PLEP-0008.rst>`__                                        |                                                                    |
```
Do we have a DOI to put here now?

rocco8773 commented on PLEP-0000.rst:
```diff
@@ -2,81 +2,116 @@
++--------+--------------------------------------------------------------+----------------------+------------+---------------------------+
+| 6      | `A New General-Purpose Plasma Object`_                       | Andrew J. Leonard    | standard   | `10.5281/zenodo.1460976`_ |
++--------+--------------------------------------------------------------+----------------------+------------+---------------------------+
+| 7      | `Structure of Top-Level Sub-Packages`_                       | Erik T. Everson,     | standard   |                           |
```
This is missing its DOI.

namurphy replied:
This hasn't been uploaded to Zenodo, so it does not yet have a DOI.

namurphy replied:
I lied. It does have one, which I'll add.

 - [ ] Do ↑
